netconfig/doc/netconfig.5
.TH netconfig 5 "August 5, 2006" "Frugalware 0.6" ""
.SH NAME
/etc/sysconfig/network/ \- configuration of network interfaces
.SH CONFIGURATION
Every file in /etc/sysconfig/network represents a network profile. When
netconfig invoked with the start paramter, then the profile named 'default'
(/etc/sysconfig/network/default) will be read.

This behaviour can be overwritten by the "network=<profilename>" kernel
parameter. In that case the given profile will be loaded instead of 'default'.

This profile is divided into sections or \fIinterfaces\fP.  Each section
defines an interface that netconfig should initialize. The exception to this is
the \fIoptions\fP section, which defines global options.
.TP
.SH Example for static IP:
.RS
.nf
[options]
dns = 192.168.1.1

[eth0]
options = 192.168.1.102 netmask 255.255.255.0
gateway = default gw 192.168.1.1
.fi
.TP
.SH Example for static ADSL with NAT:
.RS
.nf
[options]
adsl_username = user@provider.com
adsl_password = secret
adsl_interface = eth1
[eth1]
options = 192.168.1.102 netmask 255.255.255.0
post_up = iptables -t nat -A POSTROUTING -j MASQUERADE
.fi
.RE
.SH CONFIG: OPTIONS
.TP
.B "dns = <ip>"
Explicitly defines dns addresses (ie. when using a static ip). Multiple usage is ok.
.TP
.B "domain = <domain>"
Defines the local domain name. If defined more than once, the first will be used.
.TP
.B "desc = <description>"
A brief description of this network profile. Used by netconfig.
.TP
.B "adsl_username = <username>"
If you have an adsl (pppoe) connection, your adsl username.
.TP
.B "adsl_password = <password>"
If you have an adsl connection, your adsl password.
.TP
.B "adsl_interface = <interface>"
If you have an adsl connection, the Ethernet interface connected to the DSL modem.
.SH CONFIG: INTERFACES
Each section defines an interface.
The interface name is defined by the string within square brackets (eg, the two
above are 'eth0' and 'eth1'). The available directives:
.TP
.B "options = <ip> [netmask <netmask> [broadcast <broadcast>]"
Defines the ip address, netmask and broadcast address of the interface.
Multiple usage is ok for eth0:1, eth0:2, etc. See the other format of this
directive for dhcp.
.TP
.B "options = dhcp"
Use dhcp for this interface. When this format is used, the other "options" lines are ignored.
.TP
.B "pre_up = <command>"
.TP
.B "post_up = <command>"
.TP
.B "pre_down = <command>"
.TP
.B "post_down = <command>"
The syntax of these directives are the same: you can define there commands to be executed
before the interface is up, after the interface is up, before it is down, and after,
respectively. Multiple usage is ok, the commands are executed in the specified order.
.TP
.B "mac = <mac address>"
Overwrites the mac address of the card. If defined more than once, the first will be used.
.TP
.B "essid = <essid>"
Set the ESSID (or Network Name - in some products it may also be called Domain ID) for
wireless network adapters. If defined more than once, the first will be used.
.TP
.B "mode = <mode>"
Set the operating mode of the device, for wireless network adapters. The mode
can be  Ad- Hoc  (network  composed  of  only one cell and without Access
Point), Managed (node connects to a network composed of many Access Points,
with roaming), Master (the node  is  the  synchronisation master  or  acts  as
an Access Point), Repeater (the node forwards packets between other wireless
nodes), Secondary (the node acts as a backup master/repeater), Monitor (the
node is  not  associ- ated with any cell and passively monitor all packets on
the frequency) or Auto. If defined more than once, the first will be used.
.TP
.B "gateway = [default gw] <ip>"
Route packets via a gateway. NOTE: The specified gateway must be reachable first. If the
"default gw" prefix is used, then adds a default route. If defined more than once,
the first will be used. If using dhcp, then this directive is ignored.
.TP
.B "key = <key>"
Used to manipulate encryption or scrambling keys and security mode for wireless networks.
Examples:
.RS
.nf
key = 0123-4567-89
key = [3] 0123-4567-89
key = s:password [2]
key = [2]
key = open
key = off
key = restricted [3] 0123456789
key = 01-23 key 45-67 [4] key [4]
.fi
.RE
.TP
.B "dhcp_opts = <options>"
These options are passed to the dhcp client daemon. The default is "-t 10". You may want to
specify here the dhcp hostname, then you'll need "-t 10 -h <hostname>".
.TP
.B "dhcpclient = <dhcp client>"
This specifies the dhcp client to use. It can be either dhcpcd (default) or dhclient.
.SH "SPECIAL WIRELESS CARDS"
Some users reported that they had to initialize the card with ifconfig
<interface> up before using iwconfig.

Example:
.nf
[eth1]
pre_up = ifconfig eth1 up
post_down = ifconfig eth1 down
.fi

NOTE: if you use several pre_up and/or post_down lines, this pre_up should be
the first, and this post_down should be the last one.

.SH "USING WPA_SUPPLICANT"

If you want to use wpa_supplicant instead of the simple encryption (using key =
secret), then use the following pre_up and post_down lines:

.nf
pre_up = wpa_supplicant -Dwext -iwlan0 -c /etc/wpa_supplicant.conf -w -B
post_down = killall wpa_supplicant
.fi

You also need to edit /etc/wpa_supplicant.conf: comment every network block and
add one like this:

.nf
network={
        ssid="youressid"
        psk="secret"
}
.fi

.SH AUTHORS
Written by Miklos Vajna.
.SH "REPORTING BUGS"
If you find any bug, then please file a bugreport at <http://bugs.frugalware.org/>.
.SH "SEE ALSO"
.BR ifconfig (8),
.BR route (8),
.BR iwconfig (8),
.BR dhcpcd (8)


NEWS
VERSION         DESCRIPTION
-----------------------------------------------------------------------------
0.7.5         - fixed to work with os-prober >= 1.21
              - new Danish translation
0.7.4         - grubconfig, libfrugalwareutils: spelling fixes
              - added Italian translation
              - converted to git
0.7.3         - netconfig: tons of minos bugfixes and improvements
	      - grubconfig: support for kernel parameters containing double
	        dots
              - timeconfig: allow language-specific default timezones
              - xconfig: spelling fixes
0.7.2         - configure now checks for all dependencies
	      - netconfig: command line option to specify a specific interface
	        to start/stop
              - dhcpcd lockfiles moved from /etc/dhcpc to /var/run
              - grubconfig: fix menu.lst for raid1 devices
0.7.1         - minor improvements in timeconfig
              - tons of bugfixes in libfwnetconfig
              - netconfig: new --loup and --lodown switches
              - new Swedish translation
0.7.0         - new Slovak translation
              - fixed some memory leak in libfwnetconfig
              - documentation fixes
              - fix for netconfig stop when using dhcp
0.6.9         - netconfig:
                - don't ask if the user wants a dsl conn in dsl mode
                - grammar fixes
                - prevent multiple declaration of libfwnetconfig.h
              - ported to libpacman
0.6.8         - netconfig: handle SIGTERM
              - grubconfig: handle the case when there is no message file
              - build fix for newer intltool
              - moved the po files to the tranlations repo
0.6.7         - netconfig: new directive in the config file: mode
              - merged documentation updates from the wiki
              - updated Hungarian translation
0.6.6         - netconfig reverted the dhclient update since it does not have
                any switch to specify the dhcp hostname
              - added Polish translation
              - updated French/Hungarian translation
              - netconfig: new option --loop
0.6.5         - netconfig: the default dhcp client is dhclient now
              - fix i18n build
              - fix x86_64 build
              - updated Hungarian translation
0.6.4         - now using autoconf for the buildsystem (many new switches)
              - {libfw,}netconfig: several bugfixes
0.6.3         - netconfig: custom dhcp clients / dhclient improvements
              - install header files by default
0.6.2         - new config tool: netconfig (the last one for a while)
              - libfwxconfig: replace imps2 with imps/2
              - xwmconfig: bugfix for running it for the fist time
0.6.1         - bugfixes in libfwxconfig, libfwutil
              - new config tool: xwmconfig
              - netconfig: support for interfaces != eth0 in the menu-driven
                configurator
              - libfwgrubconfig: support for newer os-prober
0.6.0         - libfwgrubconfig: support for os-prober, memtest support
              - libfwutil: descriptor leak fixes
              - grubconfig: don't init anything till the "please wait" splash
0.5.9         - new config tool: grubconfig
              - sort plugins in setup
              - raidconfig: bugfix in the Hungarian translation
              - added several French translation
0.5.8         - new config tool: xconfig
0.5.7         - bugfix release for netconfig
0.5.6         - new config tool: mouseconfig
              - fixed exiting from the setup in case the config tool would not
                exit() and the menu would return back
0.5.5         - added gettext support to raidconfig
              - fixed gettext support of timeconfig
0.5.4         - new config tool: raidconfig
              - several internal cleanups
0.5.3         - added developer documentation
0.5.2         - fix build on x86_64
              - new binary: setup - a wrapper to all config tool
0.5.1         - new config tool: timeconfig
              - new libraries: libfwutils, libfwdialog, libfwnetconfig
              - added localized manpage support to the buildsystem
0.5.0         - initial import


netconfig/netconfig.c
/*
 *  netconfig.c for frugalwareutils
 * 
 *  Copyright (c) 2006 by Miklos Vajna <vmiklos@frugalware.org>
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, 
 *  USA.
 */

#include <stdio.h>
#include <net/if.h>
#include <glib.h>
#include <libfwdialog.h>
#include <libfwutil.h>
#include <libfwnetconfig.h>
#include <setup.h>
#include <getopt.h>
#include <stdlib.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <libintl.h>

extern int fwutil_dryrun;

int nco_usage  = 0;
int nco_fast   = 0;
int nco_loup   = 0;
int nco_lodown = 0;

fwnet_profile_t *sigprof;

int usage(const char *myname)
{
	printf(_("usage: %s [options] start|stop|restart|status|list [interface]\n"), myname);
	printf(_("       %s [options] [profile] [interface]\n"), myname);
	printf(_("-h | --help              This help.\n"));
	printf(_("-f | --fast              Fast mode, used by the setup.\n"));
	printf(_("     --dry-run           Do not actually perform the operation.\n"));
	printf(_("     --loup              Bring up the loopback interface.\n"));
	printf(_("     --lodown            Bring down the loopback interface.\n"));
	return(0);
}

char *selnettype()
{
	int typenum=4;
	char *types[] =
	{
		"dhcp", _("Use a DHCP server"),
		"static", _("Use a static IP address"),
		"dsl", _("DSL connection"),
		"lo", _("No network - loopback connection only")
	};
	return(fwdialog_menu(_("Select network connection type"),
		_("Now we need to know how your machine connects to the network.\n"
		"If you have an internal network card and an assigned IP address, gateway, and DNS, use 'static' "
		"to enter these values. You may also use this if you have a DSL connection.\n"
		"If your IP address is assigned by a DHCP server (commonly used by cable modem services), select 'dhcp'. \n"
		"If you just have a network card to connect directly to a DSL modem, then select 'dsl'. \n"
		"Finally, if you do not have a network card, select the 'lo' choice. \n"),
		0, 0, 0, typenum, types));
}

int dsl_hook(fwnet_profile_t *profile, int confirm)
{
	struct stat buf;
	char *iface, *uname, *pass1, *pass2;

	// do we have pppoe?
	if(stat("/usr/sbin/pppoe", &buf))
		return(0);
	if(confirm && !fwdialog_yesno(_("DSL configuration"), _("Do you want to configure a DSL connection now?")))
		return(0);
	uname = fwdialog_ask(_("Enter user name"),
			_("Enter your PPPoE user name:"), NULL);
	snprintf(profile->adsl_username, PATH_MAX, uname);
	while(1)
	{
		pass1 = fwdialog_password(_("Password"),
				_("Please enter your PPPoE password"));
		pass2 = fwdialog_password(_("Password"),
				_("Please re-enter your PPPoE password"));
		if(!strcmp(pass1, pass2))
		{
			snprintf(profile->adsl_password, PATH_MAX, pass1);
			break;
		}
		if(!fwdialog_yesno(_("Passwords don't match"),
					_("Sorry, the passwords do not match. Try again?")))
			return(1);
	}
	iface = fwdialog_ask(_("Enter interface name"),
			_("Enter the Ethernet interface connected to the DSL modem. It will be ethn, "
				"where 'n' is a number.\n"
				"If unsure, just hit enter.\n"
				"Enter interface name:"), "eth0");
	snprintf(profile->adsl_interface, IF_NAMESIZE, iface);
	if(nco_fast)
	{
		fwutil_system("mkdir /var/run");
		fwutil_system("mount -t devpts none /dev/pts");
		fwutil_system("pppoe-connect >/dev/tty4 2>/dev/tty4 &");
		return(0);
	}
	return(0);
}

int dialog_config(int argc, char **argv)
{
	FILE *input = stdin;
	fwnet_profile_t *newprofile=NULL;
	fwnet_interface_t *newinterface = NULL;
	char option[50];
	char *ptr;
	char *host, *nettype;
	char *ipaddr=NULL, *netmask=NULL, *dns=NULL, *iface=NULL;

	dialog_state.output = stderr;
	if(argv!=NULL)
		init_dialog(input, dialog_state.output);
	fwdialog_backtitle(_("Network configuration"));

	if(!nco_fast)
		host = fwdialog_ask(_("Enter hostname"), _("We'll need the name you'd like to give your host.\n"
		"The full hostname is needed, such as:\n\n"
		"frugalware.example.net\n\n"
		"Enter full hostname:"), "frugalware.example.net");
	else
		host = strdup("frugalware.example.net");
	nettype = selnettype();

	if((newprofile = (fwnet_profile_t*)malloc(sizeof(fwnet_profile_t))) == NULL)
		return(1);
	memset(newprofile, 0, sizeof(fwnet_profile_t));
	// TODO: here the profile name ('default') is hardwired
	sprintf(newprofile->name, "default");
	if((newinterface = (fwnet_interface_t*)malloc(sizeof(fwnet_interface_t))) == NULL)
		return(1);
	memset(newinterface, 0, sizeof(fwnet_interface_t));
	if(strcmp(nettype, "lo"))
	{
		iface = fwdialog_ask(_("Enter interface name"),
		_("We'll need the name of the interface you'd like to use for your network connection.\n"
		"If unsure, just hit enter.\n"
		"Enter interface name:"), "eth0");
		snprintf(newinterface->name, IF_NAMESIZE, iface);
		newprofile->interfaces = g_list_append(newprofile->interfaces, newinterface);
	}

	if(strcmp(nettype, "lo") && fwnet_is_wireless_device(iface))
	{
		ptr = fwdialog_ask(_("Extended network name"), _("It seems that this network card has a wireless "
			"extension. In order to use it, you must set your extended netwok name (ESSID). Enter your ESSID:"),
			NULL);
		snprintf(newinterface->essid, FWNET_ESSID_MAX_SIZE, ptr);
		FWUTIL_FREE(ptr);
		ptr = fwdialog_ask(_("Encryption key"), _("If you have an encryption key, then please enter it below.\n"
			"Examples: 4567-89AB-CD or  s:password"), NULL);
		snprintf(newinterface->key, FWNET_ENCODING_TOKEN_MAX, ptr);
		FWUTIL_FREE(ptr);
	}
	if(!strcmp(nettype, "dhcp"))
	{
		ptr = fwdialog_ask(_("Set DHCP hostname"), _("Some network providers require that the DHCP hostname be"
			"set in order to connect. If so, they'll have assigned a hostname to your machine. If you were"
			"assigned a DHCP hostname, please enter it below. If you do not have a DHCP hostname, just"
			"hit enter."), NULL);
		if(strlen(ptr))
			snprintf(newinterface->dhcp_opts, PATH_MAX, "-t 10 -h %s\n", ptr);
		else
			newinterface->dhcp_opts[0]='\0';
		FWUTIL_FREE(ptr);
		sprintf(option, "dhcp");
		newinterface->options = g_list_append(newinterface->options, option);
	}
	else if(!strcmp(nettype, "static"))
	{
		// options = ip netmask netmask
		ipaddr = fwdialog_ask(_("Enter ip address"), _("Enter your IP address for the local machine."), NULL);
		netmask = fwdialog_ask(_("Enter netmask for local network"),
			_("Enter your netmask. This will generally look something like this: 255.255.255.0\n"
			"If unsure, just hit enter."), "255.255.255.0");
		if(strlen(ipaddr))
			snprintf(option, 49, "%s netmask %s", ipaddr, netmask);
		newinterface->options = g_list_append(newinterface->options, option);
		FWUTIL_FREE(ipaddr);
		FWUTIL_FREE(netmask);
		ptr = fwdialog_ask(_("Enter gateway address"), _("Enter the address for the gateway on your network. "
			"If you don't have a gateway on your network just hit enter, without entering any ip address."),
			NULL);
		if(strlen(ptr))
			snprintf(newinterface->gateway, FWNET_GW_MAX_SIZE, "default gw %s", ptr);
		FWUTIL_FREE(ptr);
		dns = fwdialog_ask(_("Select nameserver"), _("Please give the IP address of the name server to use. You can"
			"add more Domain Name Servers later by editing /etc/sysconfig/network/default.\n"
			"If you don't have a name server on your network just hit enter, without entering any ip address."),
			NULL);
		newprofile->dnses = g_list_append(newprofile->dnses, dns);
	}
	if(!strcmp(nettype, "static"))
		dsl_hook(newprofile, 1);
	if(!strcmp(nettype, "dsl"))
		dsl_hook(newprofile, 0);

	if(fwdialog_yesno(_("Adjust configuration files"), _("Accept these settings and adjust configuration files?"))
		&& !fwutil_dryrun)
		fwnet_writeconfig(newprofile, host);

	g_list_free(newinterface->options);
	FWUTIL_FREE(newinterface);
	g_list_free(newprofile->interfaces);
	FWUTIL_FREE(newprofile);
	FWUTIL_FREE(host);
	FWUTIL_FREE(nettype);
	FWUTIL_FREE(ipaddr);
	FWUTIL_FREE(netmask);
	FWUTIL_FREE(dns);
	if(argv!=NULL)
		end_dialog();
	return(0);
}

int run(int argc, char **argv)
{
	int opt;
	int option_index;
	static struct option opts[] =
	{
		{"help",           no_argument,       0, 'h'},
		{"fast",           no_argument,       0, 'f'},
		{"dry-run",        no_argument,       0, 1000},
		{"loup",           no_argument,       0, 1001},
		{"lodown",         no_argument,       0, 1002},
		{0, 0, 0, 0}
	};
	char *fn=NULL, *iface = NULL;
	int ret=0, i;
	fwnet_profile_t *profile;

	while((opt = getopt_long(argc, argv, "hf", opts, &option_index)))
	{
		if(opt < 0)
			break;
		switch(opt)
		{
			case 1000: fwutil_dryrun = 1; break;
			case 1001: nco_loup   = 1; break;
			case 1002: nco_lodown = 1; break;
			case 'h':  nco_usage  = 1; break;
			case 'f':  nco_fast   = 1; break;
		}
	}
	fwutil_i18ninit(__FILE__);
	if(nco_usage)
	{
		usage(argv[0]);
		return(0);
	}
	
	if(nco_loup)
	{
		fwnet_loup();
		return(0);
	}
	
	if(nco_lodown)
	{
		fwnet_lodown();
		return(0);
	}
	
	if(optind < argc)
	{
		if(optind + 1 < argc)
		{
			iface = argv[optind+1];
		}
		if(!strcmp("list", argv[optind]))
		{
			fwnet_listprofiles();
			return(0);
		}
		if((fn=fwnet_lastprofile()) || !strcmp("stop", argv[optind]) || !strcmp("status", argv[optind]))
		{
			if((!strcmp("stop", argv[optind])) && !fn)
				return(127);
			if((!strcmp("status", argv[optind])) && !fn)
			{
				printf(_("No profile loaded.\n"));
				return(1);
			}
			else if ((!strcmp("status", argv[optind])) && fn)
			{
				printf(_("Current profile: %s\n"), fn);
				return(0);
			}
			profile = fwnet_parseprofile(fn);
			if(profile!=NULL)
				// unload the old profile
				for (i=0; i<g_list_length(profile->interfaces); i++)
				{
					fwnet_interface_t* j = g_list_nth_data(profile->interfaces, i);
					if(!iface || !strcmp(iface, j->name))
						fwnet_ifdown(j, profile);
				}
			if(!strcmp("stop", argv[optind]))
			{
				if(!fwutil_dryrun)
					fwnet_setlastprofile(NULL);
				return(0);
			}
		}
		// load the default for 'start' and for 'restart' if not yet started
		if(!strcmp("start", argv[optind]) || (!strcmp("restart", argv[optind]) && !fn))
			fn = strdup("default");
		// load the target profile if != 'restart'
		else if (strcmp("restart", argv[optind]))
			fn = strdup(argv[optind]);
		// load the new profile
		profile = fwnet_parseprofile(fn);
		if(profile==NULL)
			return(1);
		for (i=0; i<g_list_length(profile->interfaces); i++)
		{
			fwnet_interface_t* j = g_list_nth_data(profile->interfaces, i);
			if(!iface || !strcmp(iface, j->name))
				ret += fwnet_ifup(j, profile);
		}
		fwnet_setdns(profile);
		if(!fwutil_dryrun)
			fwnet_setlastprofile(fn);
		FWUTIL_FREE(fn);
	}
	else
		dialog_config(argc, argv);
	return(ret);
}

plugin_t plugin =
{
	"netconfig",
	"Network configuration",
	run,
	NULL // dlopen handle
};

plugin_t *info()
{
	return &plugin;
}
